Added additional data to the TeX export file.

mccand opened the pull request:
Previously, the exporting to TeX generated a document that contained
very limited information about the dive.  For example: the only details
reported about air consumption were delta pressure and SAC.  In addition
no data was reported about what equipment (weights, tanks, suit) was
used.
    
This update adds the following information:
   * GPS Location
   * Dive computer used
   * Max, Min, Air, and Water Temperatures
   * Average depth
   * Number of cylinders used, gas mix, and start/end pressure for each
   * Total weight used, and the description and quantity of each weight
   * Dive rating (previously, only visibility was reported)
   * Dive master
   * Suit

This update removed several fields that had no output. (i.e. "spot")
    
The purpose of this work is to allow for generation of more complete log
pages using TeX/LaTeX.
    
Future work:
   * There should be a setting allowing all units to be set to either
     Metric or imperial.
   * The export currently only includes the first entry in the tag list.
     Future versions should export all tags.
   * Ideally, this would export a screenshot of the dive location.
   * It would be nice to have a bit more control over the generation of
     profile images, but that may be beyond me.

Finally, this is all new to me (first commit to a project) so I apologize
in advance for any missteps.
    
Signed-off-by: Ben McCandless <mccand@gmail.com>

atdotde commented on desktop-widgets/divelogexportdialog.cpp:
```diff
@@ -275,31 +274,77 @@ void DiveLogExportDialog::export_TeX(const char *filename, const bool selected_o
+		qty_cyl = 0;
+		for (i = 0; i < MAX_CYLINDERS; i++)
+			if (is_cylinder_used(dive, i)){
+				put_format(&buf, "\\def\\cyl%ddescription{%s}\n", i + 1, dive->cylinder[i].type.description);
```
I don't see how this will compile in TeX: Remember, you need a stylefile as well. In order to print those fields, you need to mention the macros like \cylidner23description somewhere. If in that particular dive, cylinder 23 is unused, TeX will complain about an undefined macro. Better: If unused define to be empty or \relax. And for TeX, the actual cylinder number probably does not matter, so better do an extra counter and fill up the list without unused cylinders.

mccand replied:
As embarrassed as I am to admit it, you're right!  This doesn't compile - I should have had a \% rather than just % in the output.  I'll get that fixed.  However, once I inserted that backslash and checked the template for untranslated keywords (\klein = \small), it processed fine - though it's not pretty.

There is one change in that I renamed /place as /sitename. (The template would need that one line to be updated.)

Here is a sample output file produced by this code:
`\def\cyl1description{AL80}
\def\cyl1mixO2{21%}  \\note should read "...{21\%}" in fixed version.  Same for all fields below.
\def\cyl1mixN2{79%}
\def\cyl1mixHe{0%}
\def\cyl1startpress{213.7 bar}
\def\cyl1endpress{34.4 bar}
\def\cyl2description{AL100}
\def\cyl2mixO2{40%}
\def\cyl2mixN2{0%}
\def\cyl2mixHe{60%}
\def\cyl2startpress{200.0 bar}
\def\cyl2endpress{20.0 bar}
\def\qtycyl{2}`

So far, I have been diving single tanks exclusively, so all I really need to print is the data for cyl1.

I'll take a look at the case where cylinder #3 is used, but cylinder #2 is not, and see if I can get to work out nicely.  For my implementation, I want to be able to print the cylinders that were used, and I don't care about the ones that were not.

atdotde replied:
In that case, in addition to i use a separate counter that counts the used cylinders and use that for the TeX macros. And explicitly define the unused ones to \relax.

atdotde commented on desktop-widgets/divelogexportdialog.cpp:
```diff
@@ -275,31 +274,77 @@ void DiveLogExportDialog::export_TeX(const char *filename, const bool selected_o
+		for (i = 0; i < MAX_CYLINDERS; i++)
+			if (is_cylinder_used(dive, i)){
+				put_format(&buf, "\\def\\cyl%ddescription{%s}\n", i + 1, dive->cylinder[i].type.description);
+				if (dive->cylinder[i].gasmix.o2.permille > 0){ //This code assumes that all gas must have oxygen, so a zero value indicated that the gas in question is AIR
```
No no no. Don't access .permille directly. We have helper functions for that like get_o2(). Those can deal with the air special case and there are also functions that provide gas names.

mccand replied:
Forgive me, but I'm utterly new at this.  I would love to have some good pointers about better ways to do things, and if there is a function that I don't know about that does something better then what I've done, I'd love to know.  As it is, I'm puzzling things out as I can, based on the documentation and code available to me.  As it is, I did my best to access variables in a similar way to that in which it had been done in other places in the current document.  I may need a bit of time to catch up to those of you who have been doing this for a while.

atdotde replied:
Have a look at what is in dive.c and dive.h. There are all kinds of handy helper functions. Here you probably want to use gassname(const struct *gasmix) defined in equipment.c If you really want to know the O2 fraction, use get_o2(), there is also get_he(). To print unitful quantities, you can use get_depth_units(int mm, int *frac, const char **units) and similar.

atdotde commented on desktop-widgets/divelogexportdialog.cpp:
```diff
@@ -275,31 +274,77 @@ void DiveLogExportDialog::export_TeX(const char *filename, const bool selected_o
+					put_format(&buf, "\\def\\cyl%dmixHe{0\%}\n", i + 1);
+				}
+				delta_p.mbar += dive->cylinder[i].start.mbar - dive->cylinder[i].end.mbar;
+				put_format(&buf, "\\def\\cyl%dstartpress{%u.%01u bar}\n", i + 1, FRACTION(dive->cylinder[i].start.mbar / 100, 10));
```
as above. What if undefined? 

An idea would be to compile this data into a TeX table but leave the template line out for the style file to supply it, like

EAN32&200&bar&72&bar&2300&l\cr
EAN50&180&bar&150&bar&550&l\cr

mccand replied:
If undefined, I ultimately want it to print a blank field (which will result in no text being printed at that location in the output file).

I'm not sure I like the idea of printing a table directly into the export file, as the benefit of using discrete fields (as opposed to writing string that defines a table directly) is that we can then define in the template what date should be printed in the final pages.  This is why I had an output counter - it will let the template figure out how big of a table should be constructed.

atdotde commented on desktop-widgets/divelogexportdialog.cpp:
```diff
@@ -275,31 +274,77 @@ void DiveLogExportDialog::export_TeX(const char *filename, const bool selected_o
+		total_weight = 0;
+		for (i = 0; i < MAX_WEIGHTSYSTEMS; i++)
+			if (dive->weightsystem[i].weight.grams != NULL){
+				put_format(&buf, "\\def\\weight%dtype{%s}\n", i + 1, dive->weightsystem[i].description);
```
Same here.

mccand replied:
Right.  same reply.

mccand commented on desktop-widgets/divelogexportdialog.cpp:
```diff
@@ -275,31 +274,77 @@ void DiveLogExportDialog::export_TeX(const char *filename, const bool selected_o
 		put_format(&buf, "\\def\\date{%04u-%02u-%02u}\n",
 		      tm.tm_year, tm.tm_mon+1, tm.tm_mday);
 		put_format(&buf, "\\def\\number{%d}\n", dive->number);
-		put_format(&buf, "\\def\\place{%s}\n", site ? site->name : "");
```
I'd be happy to leave those in.  I'll make the adjustment.

mccand commented:
Recap: I need to fix a few things:
- Put legacy fields back in.
- add backslash in percentage
- If I can get some pointers on other functions to use, update my code to remove some of the special cases.

A question for anyone able to answer it: I'm still learning a lot about how github works - do I need to cancel this pull request and put up a new one after changes are implemented, or will they be rolled into this request?

Thanks,

Ben

atdotde commented:
One last bullet point: If you define a macro in some cases (cylinder 23 is used) make sure it is always defined (possibly with an empty default).

You can modify your commit (which is probably what I would do in your case) using git reset or git commit --amend and then push it to GitHub with --force. This will update the pull request automatically. Or you add more commits to this branch. Those will as well automatically go into the pull request.

dirkhh commented:
What's the state of this? @atdotde clearly had comments / concerns. @mccand are you planning to update the pull request? Or is this abandoned and should be closed?

mccand commented:
DIrk,

I'm working on addressing the comments, but have been on vacation in the
Yukon - away from my computer most of the time.  I'm back in regular
civilization for now, and will try to make some progress in the next couple
of days.

That said - I'm, uh, "woefully unexperienced" in programming with C++, so
it's been a slow process for me.

Ben

On Sun, Jul 23, 2017 at 5:04 PM Dirk Hohndel <notifications@github.com>
wrote:

> What's the state of this? @atdotde <https://github.com/atdotde> clearly
> had comments / concerns. @mccand <https://github.com/mccand> are you
> planning to update the pull request? Or is this abandoned and should be
> closed?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/Subsurface-divelog/subsurface/pull/477#issuecomment-317295890>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABwhG_GyvqL-DRnjdeG6vwr7pmRHRFkoks5sQ-2BgaJpZM4OUkQ0>
> .
>

dirkhh commented:
@mccand 
Ben, no problem - there's no rush, I just wanted to know what the status was. Thanks for the update.

mccand commented:
Okay, I think I have addressed all the comments, and fixed some of the other things that were bothering me too.  I'm going to attempt to update this pull request and see if it will bring through the newest code.

Ben

janmulder commented:
Some remarks:
- `\def\gpslat` and `gpslon` are double defined in the output (due to faulty merge of the conflict in 45587c4). Does not seem to hurt but seems easy to correct.
- I do not understand why the TeX output starts with a set of defines for units, but those are not used later on.
- again a lot of double defines in the output (Dive profile Details), for example `maxtemp`, `mintemp` etc.
- `\def\cyl1mixO2{10%}`is missing a `\` for the `%` (this is correct later on in the TeX output). Obviously, breaking the parsing from TeX.
- After fixing the missing `\`, TeX parsing still does not work for me. 

```
! Illegal unit of measure (pt inserted).
<to be read again> 
                   m
<to be read again> 
                   \dimen@ 
```
As I'm a little rusty on my plain TeX skills, I do not see the error in the output right away.

- Finally, I think Dirk would appreciate the pull request to be rebased on the current master ....

dirkhh commented:
Sorry for the top post. On vacation in London 😀
No need to rebase, I'll happily take care of that. But I'd like to see an ack from @atdotde



-------- Original Message --------
From: janmulder <notifications@github.com>
Sent: Sun Aug 06 10:09:35 GMT+01:00 2017
To: Subsurface-divelog/subsurface <subsurface@noreply.github.com>
Cc: Dirk Hohndel <dirk@hohndel.org>, Comment <comment@noreply.github.com>
Subject: Re: [Subsurface-divelog/subsurface] Added additional data to the TeX export file. (#477)

Some remarks:
- `\def\gpslat` and `gpslon` are double defined in the output (due to faulty merge of the conflict in 45587c4). Does not seem to hurt but seems easy to correct.
- I do not understand why the TeX output starts with a set of defines for units, but those are not used later on.
- again a lot of double defines in the output (Dive profile Details), for example `maxtemp`, `mintemp` etc.
- `\def\cyl1mixO2{10%}`is missing a `\` for the `%` (this is correct later on in the TeX output). Obviously, breaking the parsing from TeX.
- After fixing the missing `\`, TeX parsing still does not work for me. 

```
! Illegal unit of measure (pt inserted).
<to be read again> 
                   m
<to be read again> 
                   \dimen@ 
```
As I'm a little rusty on my plain TeX skills, I do not see the error in the output right away.

- Finally, I think Dirk would appreciate the pull request to be rebased on the current master ....




-- 
You are receiving this because you commented.
Reply to this email directly or view it on GitHub:
https://github.com/Subsurface-divelog/subsurface/pull/477#issuecomment-320495005

mccand commented:
@janmulder - Thanks for taking a look!  I've pushed fixes, I think.

- I've fixed the GPS lat/lon (Sorry, I'm still figuring out git - I feel like a kid in an airplane cockpit. "What does this lever do again?")
- The units are used later.  For example, \weightunit is seen later in the output as: 
		"\def\totalweight{5.00 \weightunit}"
The idea with defining the units up front was that it lets the user substitute other things easily, should they want to.  For example, if you want '#' instead of 'lb', or 'ℓ' instead of 'L', just change it in that one place.  
- Fixed other double-defines.
- My output has the \ preceding the %, but it looks like the conflict you mentioned above made a mess of things, and I wasn't on top of things enough to catch it.  I think it's fixed now.
- I think this was caused by overloading "\maxdepth"  I think I had this fixed before the conflict (sigh), by naming the field \maximumdepth instead.  I've run a couple of files through pdftex and now I get some warnings, but no errors.

Thanks again for taking a look at this. I'm sorry for making you point out my rookie mistakes, but _I_ really appreciate it.  

@dirkhh - based on my best "bull in a china shop" impression so far, it might be wise for you to do the rebasing.  At least until I figure out what rebasing is, and how to do it without breaking _everything_.  Enjoy your vacation!  I'll look for further comments from @atdotde 

Ben

dirkhh commented:
@mccand - no worries, I'll rebase the changes once @atdotde tells me that this is in a state where it makes sense to merge. My responses will be a bit slower because of the vacation, I'm trying to not be at the computer all the time :-)

atdotde commented:
Will be on a train later today and hopefully have some time to sort this out.

mccand commented:
Ugh.  I found an issue with the cylinder data.  I need to change the naming scheme for these fields in some way, because (La)TeX won't accept numbers in macros by default (it interprets everything after the number as unrelated text to print.  There also seems to be some differences between how LaTeX and TeX handle the same field names.  I'll try to have a solution today.

atdotde commented:
I have done some cleaning up an d rebasing of this as
https://github.com/Subsurface-divelog/subsurface/pull/537

Please use that as a base for further work.

mccand commented:
So, I'm not exactly clear on what to do here.  Do I close this pull request and start working on the new one that you've opened?

atdotde commented:
Yes, that new one is rebased on current master and eliminates the confusion from the merge mishap.

mccand commented:
Closing this in favor of #537
